src/Servers/Kestrel/Transport.NamedPipes/test/NamedPipeConnectionListenerTests.cs

```diff
@@ -163,4 +163,56 @@ public async Task BindAsync_ListenersSharePort_DisposeFirstListener_Success()
         await using var connectionListener2 = await NamedPipeTestHelpers.CreateConnectionListenerFactory(LoggerFactory, pipeName: pipeName);
         Assert.Equal(connectionListener1.EndPoint, connectionListener2.EndPoint);
     }
+
+    [Fact]
+    public async Task CompleteTransport_BeforeClientHasReadLastBytes_DontLooseData()
+    {
+        var options = new NamedPipeTransportOptions();
+        await using var connectionListener = await NamedPipeTestHelpers.CreateConnectionListenerFactory(LoggerFactory, options: options);
+
+        var clientStream = NamedPipeTestHelpers.CreateClientStream(connectionListener.EndPoint);
+        var connectTask = clientStream.ConnectAsync();
+
+        var connection = await connectionListener.AcceptAsync();
+        await connection.Transport.Output.WriteAsync(new byte[] { 41 });
+        byte[] readBuffer = new byte[1];
+        var readAmount = await clientStream.ReadAsync(readBuffer);
+        Assert.Equal(1, readAmount);
+        Assert.Equal(41, readBuffer[0]);
+        await connection.Transport.Output.WriteAsync(new byte[] { 42 });
+        // complete transport and wait for a reasonable amount of time
+        // actual stream disconnection should happen after the last client read happens
+        connection.Transport.Output.Complete();
+        await Task.Delay(500);
```

Comment:
We try hard to avoid delays in integration tests.

There are `_receivingTask` and `_sendingTask` fields in `NamedPipeConnection`. They report when the send/receive loops are done. Could one of them be made internal and then awaited by this test instead?

Reply:
The problem here, is that with WriteThrough option, the _stream.WriteAsync call in DoSendAsync do not complete until the client reads the bytes (that is why the test pass IMO). So the _sendingTask won't complete until the client has read it.

The delay here, is to make sure the current behavior stay the same (the sending task should not complete until the client has received all bytes).